server/api/impl/upstream/uphttp/upstream.go

```diff
@@ -0,0 +1,110 @@
+// Copyright (c) 2020-present Mattermost, Inc. All Rights Reserved.
+// See License for license information.
+
+package uphttp
+
+import (
+	"encoding/json"
+	"io"
+	"net/http"
+	"time"
+
+	"github.com/dgrijalva/jwt-go"
+
+	"github.com/mattermost/mattermost-plugin-apps/server/api"
+	"github.com/mattermost/mattermost-plugin-apps/server/api/impl/upstream"
+	"github.com/mattermost/mattermost-plugin-apps/server/utils/httputils"
+)
+
+type Upstream struct {
+	rootURL   string
+	appSecret string
+}
+
+func NewUpstream(app *api.App) *Upstream {
+	return &Upstream{app.Manifest.HTTPRootURL, app.Secret}
+}
+
+func (u *Upstream) InvokeNotification(n *api.Notification) error {
+	// TODO
+	resp, err := u.post("", u.rootURL+"/notify/"+string(n.Subject), n)
+	if err != nil {
+		return err
+	}
+	defer resp.Body.Close()
+	return nil
+}
+
+func (u *Upstream) InvokeCall(call *api.Call) *api.CallResponse {
+	resp, err := u.post(call.Context.ActingUserID, u.rootURL+call.URL, call)
+	if err != nil {
+		return api.NewErrorCallResponse(err)
+	}
+	defer resp.Body.Close()
+
+	cr := api.CallResponse{}
+	err = json.NewDecoder(resp.Body).Decode(&cr)
+	if err != nil {
+		return api.NewErrorCallResponse(err)
+	}
+	return &cr
+}
+
+func (u *Upstream) GetBindings(call *api.Call) ([]*api.Binding, error) {
```

Comment:
I left this in the interface because the go `json` peculiarities will not easily allow to decode specific types of `Data`, like []*Binding in this case.

The alternative was to expose the res.Body stream to the caller, and let the caller fo her own json.Decode, but I didn't like it.